include/mqtt/endpoint.hpp

```diff
@@ -8278,1217 +8319,3822 @@ class endpoint : public std::enable_shared_from_this<endpoint<Socket, Mutex, Loc
+                        );
+                    buf.remove_prefix(Bytes);
+                    handler(
+                        packet_id,
+                        std::move(buf),
+                        std::move(func),
+                        std::move(self)
+                    );
+               }
+            );
         }
-        i += client_id_length;
+    }
 
-        clean_session_ = connect_flags::has_clean_session(byte8);
-        // TODO: Implement a "will_view" which, instead of containing string objects, only
-        //       has string_views into the message buffer. This "will_view" would need to
-        //       be assigned to a real "will" object in order to have it's contents preserved.
-        mqtt::optional<will> w;
-        if (connect_flags::has_will_flag(byte8)) {
+    // This function isn't used for remaining lengh.
+    void process_variable_length(
+        async_handler_t func,
+        buffer buf,
+        std::function<void(std::size_t, buffer, async_handler_t, this_type_sp)> handler,
+        this_type_sp self
+    ) {
+        process_variable_length_impl(
+            std::move(func),
+            std::move(buf),
+            std::move(handler),
+            0,
+            1,
+            std::move(self)
+        );
+    }
 
-            std::vector<v5::property_variant> will_props;
-            if (version_ == protocol_version::v5) {
-                char const* b = payload_.data() + i;
-                char const* it = b;
-                char const* e = b + payload_.size() - i;
-                if (auto props_opt = v5::property::parse_with_length(it, e)) {
-                    will_props = std::move(*props_opt);
+    void process_variable_length_impl(
+        async_handler_t func,
+        buffer buf,
+        std::function<void(std::size_t, buffer, async_handler_t, this_type_sp)> handler,
+        std::size_t size,
+        std::size_t multiplier,
+        this_type_sp self
+    ) {
+        if (remaining_length_ == 0) {
+            call_message_size_error_handlers(func);
+            return;
+        }
+        --remaining_length_;
+
+        // I use rvalue reference paramter to reduce move constructor calling.
+        // This is a local lambda expression invoked from this function, so
+        // I can controll all callers.
+        auto proc =
+            [this]
+            (
+                async_handler_t&& func,
+                buffer&& buf,
+                std::function<void(std::size_t, buffer, async_handler_t, this_type_sp)>&& handler,
+                std::size_t size,
+                std::size_t multiplier,
+                this_type_sp&& self
+            ) mutable {
+                if (!calc_variable_length(size, multiplier, buf.front())) {
+                    call_message_size_error_handlers(func);
+                    return;
+                }
+                if (buf.front() & 0b10000000) {
```

Comment:
I think that 0b10000000 should be a named constant.

E.g.

```cpp
static constexpr std::uint8_t variable_length_flag = 0b10000000
```

Reply:
Ok I will introduce the constant.